internal/controller/maintenancemode_controller.go

```diff
@@ -0,0 +1,202 @@
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	maintenanceModeFinalizer = "ocs-client-operator.ocs.openshift.io/maintenance-mode"
+	ramenReplicationIDLabel  = "ramendr.openshift.io/replicationID"
+	MaintenanceModeCRDName   = "maintenancemodes.ramendr.openshift.io"
+)
+
+// MaintenanceModeReconciler reconciles a ClusterVersion object
+type MaintenanceModeReconciler struct {
+	client.Client
+	OperatorNamespace string
+	Scheme            *runtime.Scheme
+
+	log logr.Logger
+	ctx context.Context
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *MaintenanceModeReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	generationChangePredicate := predicate.GenerationChangedPredicate{}
+	bldr := ctrl.NewControllerManagedBy(mgr).
+		For(&ramenv1alpha1.MaintenanceMode{}, builder.WithPredicates(generationChangePredicate)).
+		Watches(&storagev1.StorageClass{}, &handler.EnqueueRequestForObject{}, builder.WithPredicates(generationChangePredicate)).
+		Watches(&v1alpha1.StorageClient{}, &handler.EnqueueRequestForObject{}, builder.WithPredicates(generationChangePredicate))
+
+	return bldr.Complete(r)
+}
+
+//+kubebuilder:rbac:groups=ramendr.openshift.io,resources=maintenancemodes,verbs=get;list;update;create;watch;delete
+//+kubebuilder:rbac:groups=ramendr.openshift.io,resources=maintenancemodes/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=ramendr.openshift.io,resources=maintenancemodes/finalizers,verbs=update
+//+kubebuilder:rbac:groups=ocs.openshift.io,resources=storageclients,verbs=get;list;watch
+//+kubebuilder:rbac:groups=storage.k8s.io,resources=storageclasses,verbs=get;list;watch
+
+func (r *MaintenanceModeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.ctx = ctx
+	r.log = log.FromContext(ctx, "MaintenanceMode", req)
+	r.log.Info("Reconciling MaintenanceMode")
+
+	maintenanceMode := &ramenv1alpha1.MaintenanceMode{}
+	maintenanceMode.Name = req.Name
+	if err := r.get(maintenanceMode); err != nil {
+		if kerrors.IsNotFound(err) {
+			r.log.Info("Maintenance Mode resource not found. Ignoring since object might be deleted.")
+			return reconcile.Result{}, nil
+		}
+		r.log.Error(err, "failed to get the Maintenance Mode")
+		return reconcile.Result{}, err
+	}
+
+	// find the storageClass with the targetID
+	storageClassList := &storagev1.StorageClassList{}
+	err := r.list(storageClassList, client.MatchingLabels{ramenReplicationIDLabel: maintenanceMode.Spec.TargetID})
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if len(storageClassList.Items) != 1 {
+		return ctrl.Result{}, fmt.Errorf("failed to find storageClass for maintenance mode")
+	}
+
+	storageClass := &storageClassList.Items[0]
+
+	// find the storageClient
+	// TODO: when storageClass is owned by storageClient, move to find the Client by ownerRef
+	storageClient := &v1alpha1.StorageClient{}
+	storageClient.Name = storageClass.GetLabels()[storageClientLabel]
```

Comment:
Please use the owner ref, i am not sure about if this label even exists anymore and we have no gurentees we will keep it in the future

Reply:
There is no onwerRef for storageClient on the StorageClasses

Reply:
Yes, I have added a TODO for it, and will open an issue for it when we move to direct ownership of storageClass by the clients

Reply:
ok, resolving